[platform] Update mellanox platform files to support updating multiple components simultaneously

alexrallen opened the pull request:
Support requirements from Microsoft to separate install process from reboot for firmwares in order to support automatic firmware update feature. 

Depends on https://github.com/alexrallen/sonic-platform-common/pull/4 AND https://github.com/alexrallen/sonic-buildimage/pull/23

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -628,7 +635,7 @@ def get_firmware_update_notification(self, image_path):
 
     def install_firmware(self, image_path):
-        return self.__install_firmware(image_path)
+        return self.__install_firmware(image_path, allow_reboot=False)
```
@alexrallen this will break the original `non-autoupdate` behaviour

alexrallen replied:
Lets merge this discussion with your comment in the upgrade script.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -649,7 +656,7 @@ def __init__(self):
         self.onie_updater = ONIEUpdater()
 
-    def __install_firmware(self, image_path):
+    def __install_firmware(self, image_path, reboot=False):
```
@alexrallen this will break the original `non-autoupdate` behaviour

alexrallen replied:
Lets merge all your concerns about the original non-auto update behavior into this thread since essentially all other firmwares other than SSD have similar requirements. 

How exactly do you think this will break the behavior? I recognize that the "install" command will no longer automatically reboot the switch but I believe this is correct with regard to the HLD.

From discussions with MSFT they claim that "install" is only performing the actions up to any reboot behavior while "update" is performing all actions necessary to fully update the firmware (including a reboot if necessary). I have updated the flows to be compatible with this definition which is also a requirement for having the auto update flow work as it assumes our API implementation acts in this way.

nazariig commented on platform/mellanox/mlnx-ssd-fw-update.sh:
```diff
@@ -178,6 +179,10 @@ function check_usage() {
             ARG_POWER_CYCLE_FLAG=$TRUE
             shift # past argument
             ;;
+	--no-power-cycle)
```
@alexrallen this will break `Innodisk` FW update. Please sync with BSP for more details

alexrallen replied:
All this does is suppress the *automatic* power cycle. With the way this is currently configured the auto update flow will still power cycle the machine immediately after installation (perhaps after finishing installing another firmware). 

The power cycle performed here is the same one performed by the register by the platform specific reboot plugin.

alexrallen replied:
The user is given a message to power cycle the box immediately if this is called from the firmware install flow.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -532,11 +538,12 @@ def auto_update_firmware(self, image_path, boot_action):
-                return FW_AUTO_INSTALLED
-            # Failed update (unknown reason)
-            return FW_AUTO_ERR_UKNOWN
+            if boot_action == 'cold':
```
@alexrallen suggest moving supported `boot_action` validation on top of this method

alexrallen replied:
Can you please clarify this suggestion? I am not sure what you mean.

nazariig replied:
@alexrallen the validation of `boot_action` can be done on top in this method: if it doesn't fit the criteria - return immediately. In general it helps to avoid a nested if-else inside

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -438,13 +440,13 @@ def __init__(self):
         self.onie_updater = ONIEUpdater()
 
-    def __install_firmware(self, image_path):
+    def __install_firmware(self, image_path, reboot=False):
```
@alexrallen this will break the original `non-autoupdate` behaviour

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -792,7 +799,17 @@ def get_firmware_update_notification(self, image_path):
 
     def install_firmware(self, image_path):
-        return self.__install_firmware(image_path)
+        if ".mpfa" in image_path:
```
@alexrallen suggest having a constant for `.mpfa`

alexrallen replied:
Agree

nazariig commented on platform/mellanox/install-pending-fw.py:
```diff
@@ -45,7 +45,7 @@
             CPLD_FLAG = (component, fw_file)
         else:
-            component.install_firmware(fw_file)
+            component.__install_firmware(fw_file, allow_reboot=False)
```
@alexrallen in python double underscore notation `__` is for private calls only. You can't use it for public APIs.

alexrallen replied:
How do you suggest I handle this?

nazariig replied:
@alexrallen use notation without underscore and define a separate API

alexrallen replied:
Sure.

nazariig commented on platform/mellanox/install-pending-fw.py:
```diff
@@ -0,0 +1,53 @@
+    exit(0)
+
+# Parse platform components file
+pcp.parse_platform_components()
```
@alexrallen how do you expect to handle errors here?

alexrallen replied:
I can bail out of the script with an error message if it throws an exception. Is there an exception we particularly need to catch?

nazariig replied:
@alexrallen so the idea is to deliver a raw trace back to the user?

alexrallen replied:
Not a raw trace, I will catch the error and output a message to the log. The question is what error message do you think I should give and is there a particular python exception I should catch?

i.e. do we anticipate this failing in a specific way? Just an ill-formated `platform_components.json` file?

nazariig replied:
> i.e. do we anticipate this failing in a specific way? Just an ill-formated platform_components.json file?
@alexrallen why not?

alexrallen replied:
Ok I will catch a broad Exception and then report back that parsing of the platform_components.json file failed and will also report the error message from python. Sound good?

The script will also bail at this point. I will add a sleep so someone can catch the error before reboot.

nazariig replied:
@alexrallen ok

nazariig commented on platform/mellanox/mlnx-onie-fw-update.sh:
```diff
@@ -186,7 +186,12 @@ case "${cmd}" in
-                system_reboot
+                if [[ "${arg}" == "--no-reboot" ]]; then
+                    echo "ONIE firmware update successfully STAGED for install at NEXT reboot. Please reboot manually to complete installation."
+		    exit 0
```
@alexrallen a typo?

alexrallen replied:
I don't see a typo, can you point it out?

nazariig replied:
@alexrallen `exit 0` should be aligned with `echo ...`?

alexrallen replied:
Oh I see, think thats a tab / space thing in my editor will fix.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -182,8 +184,12 @@ def __unstage_update(self, image_path):
-    def __trigger_update(self):
-        cmd = self.ONIE_FW_UPDATE_CMD_UPDATE
+    def __trigger_update(self, reboot):
+
```
@alexrallen what is the motivation for extra line?

alexrallen replied:
None, I can remove.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -182,8 +184,12 @@ def __unstage_update(self, image_path):
+        if reboot:
+            cmd = self.ONIE_FW_UPDATE_CMD_UPDATE
+        else:
+            cmd = self.ONIE_FW_UPDATE_CMD_INSTALL
```
@alexrallen FYI: in case a several FW binaries are staged, ONiE FW Framework can update only one at time. So eventually it will require a few cold reboots

alexrallen replied:
Understood.

alexrallen commented:
@nazariig I believe all of your feedback has been addressed. Please take a look. Thank you!

nazariig commented on platform/mellanox/install-pending-fw.py:
```diff
@@ -0,0 +1,59 @@
+from fwutil.lib import ComponentStatusProvider, PlatformComponentsParser
+from sonic_platform.component import ComponentCPLD
+
+#Globals
```
```suggestion
# Globals
```

nazariig commented on platform/mellanox/install-pending-fw.py:
```diff
@@ -0,0 +1,59 @@
+
+# Run CPLD refresh last if needed
+if CPLD_FLAG is not None:
+    CPLD_FLAG[0].__refresh_cpld(CPLD_FLAG[1])
```
@alexrallen in python you can't use private notation `__` for public API calls

nazariig commented on platform/mellanox/install-pending-fw.py:
```diff
@@ -0,0 +1,59 @@
+for boot_type, components in update_status.items():
+    for comp in components:
+
+        if comp["info"] != "scheduled": continue
```
```suggestion
        if comp['info'] != 'scheduled': continue
```

nazariig replied:
@alexrallen suggest to have a constant for `scheduled`

nazariig commented on platform/mellanox/install-pending-fw.py:
```diff
@@ -0,0 +1,59 @@
+
+        key = comp['comp']
+
+        if len(key.split('/')) == 3:
```
@alexrallen how about `csp.is_modular_chassis()`?

nazariig commented on platform/mellanox/mlnx-onie-fw-update.sh:
```diff
@@ -186,7 +186,12 @@ case "${cmd}" in
             if [[ ${rc} -eq 0 ]]; then
-                system_reboot
+                if [[ "${arg}" == "--no-reboot" ]]; then
+                    echo "ONIE firmware update successfully STAGED for install at NEXT reboot. Please reboot manually to complete installation."
```
```suggestion
                    echo "INFO: ONIE firmware update successfully STAGED for install at NEXT reboot. Please reboot manually to complete installation."
```

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -464,11 +466,14 @@ def get_available_firmware_version(self, image_path):
 
     def update_firmware(self, image_path):
-        self.__install_firmware(image_path)
+        self.__install_firmware(image_path, reboot=True)
```
@alexrallen any reason to specify the reboot explicitly?

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -299,7 +304,7 @@ def get_onie_firmware_info(self, image_path):
         return firmware_info
 
-    def update_firmware(self, image_path):
+    def update_firmware(self, image_path, reboot):
```
@alexrallen suggest to define a separate API `install_firmware_no_reboot`

alexrallen replied:
After reviewing this and what would be the cleanest solution with the least impact is to update the "install firmware" (for the components) and the "update firmware" (for the ONIEUpdater class) methods to include named parameters for specifying reboots with defaults set for the original behavior....

So in this case (ONIEUpdater) the function signature would look like....
`def update_firmware(self, image_path, allow_reboot=True)`

and in the case of components (ComponentSSD / ComponentONIE / etc.) the signature would look like...
`def install_firmware(self, image_path, allow_reboot=True)`

This will allow these methods to: 
* Comply with the abstract API defined by the SONiC platform API spec
* Perverse 100% backwards compatibility
* Allows calling them from new contexts that don't support immediate reboots

Without me having to define new methods which are extremely simply duplicates of the existing methods which I feel would adversely impact the maintainability and readability of the code.

nazariig replied:
> So in this case (ONIEUpdater) the function signature would look like....
def update_firmware(self, image_path, allow_reboot=True)

@alexrallen looks like the signature within the code is different ...

alexrallen replied:
Updated to be consistent

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -532,11 +541,12 @@ def auto_update_firmware(self, image_path, boot_action):
-                return FW_AUTO_INSTALLED
-            # Failed update (unknown reason)
-            return FW_AUTO_ERR_UKNOWN
+            if supported_boot == ['cold']:
```
```suggestion
            if boot_action == 'cold':
```

nazariig commented on platform/mellanox/mlnx-ssd-fw-update.sh:
```diff
@@ -683,7 +690,12 @@ elif [ $ARG_UPDATE_FLAG == $TRUE ]; then
+		    if [[ "yes" == "$power_policy" || $ARG_POWER_CYCLE_FLAG == $TRUE ]]; then
+
+                        if [[ $ARG_FORCE_NO_POWER_CYCLE_FLAG == $TRUE ]]; then
+				LOG_MSG_AND_EXIT "An IMMEDIATE power cycle is REQUIRED to upgrade the SSD. Please perform a cold reboot as soon as possible."
```
@alexrallen please speak with BSP on how to handle it properly for `Innodisk`. We do need to have a protection mechanism in case the user will invoke it manually with `--no-power-cycle`

alexrallen replied:
Mentioned this in the thread with BSP.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -792,7 +843,17 @@ def get_firmware_update_notification(self, image_path):
         return "Immediate power cycle is required to complete {} firmware update".format(self.name)
 
     def install_firmware(self, image_path):
-        return self.__install_firmware(image_path)
```
@alexrallen suggest to define a separate API `install_firmware_no_reboot`

alexrallen replied:
Will address your comments regarding a separate API for functions I am calling from the install pending script in the other thread.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -756,6 +772,41 @@ def __install_firmware(self, image_path):
+            print("INFO: Processing {} refresh file: firmware update".format(self.name))
+            self.__install_firmware(os.path.join(mpfa.get_path(), refresh_firmware))
+
+
```
@alexrallen please remove extra line

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -756,6 +772,41 @@ def __install_firmware(self, image_path):
 
         return True
 
+    def __refresh_cpld(self, image_path):
```
@alexrallen suggest having a different signature, e.g. `refresh_firmware`

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -756,6 +772,41 @@ def __install_firmware(self, image_path):
+                return FW_AUTO_ERROR_UNKNOWN
+
+        # boot_type did not match (skip)
+        return FW_AUTO_ERR_BOOT_TYPE
```
@alexrallen suggest reworking like this:
```python
        supported_boot = ['cold']

        # Verify image path exists
        if not os.path.exists(image_path):
            # Invalid image path
            return FW_AUTO_ERR_IMAGE

        if boot_action not in supported_boot:
            # boot_type did not match (skip)
            return FW_AUTO_ERR_BOOT_TYPE

        # Defer refresh to next boot
        if self.install_firmware(image_path):
            return FW_AUTO_SCHEDULED

        return FW_AUTO_ERROR_UNKNOWN
```

alexrallen replied:
Didn't end up with this exact code but I see the spirit of your comment I think. Updated these functions so they are syntactically similar to each other and easier to read / debug.

nazariig commented:
> @nazariig I believe all of your feedback has been addressed. Please take a look. Thank you!

@alexrallen please handle a new set of comments

alexrallen commented:
@nazariig Have responded to all feedback. Please take a look.

nazariig commented on platform/mellanox/install-pending-fw.py:
```diff
@@ -0,0 +1,77 @@
+        else: 
+            parent_name, comp_name = key.split('/')
+            
+        fw_file = pcp.chassis_component_map[parent_name][comp_name]["firmware"]
```
@alexrallen modular chassis platform use different component map: `module_component_map`

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -792,7 +825,17 @@ def get_firmware_update_notification(self, image_path):
+                burn_firmware = mpfa.get_metadata().get('firmware', 'burn')
+
+                print("INFO: Processing {} burn file: firmware install".format(self.name))
+                self.__install_firmware(os.path.join(mpfa.get_path(), burn_firmware))
```
@alexrallen it doesn't comply the API: missing return statement

alexrallen replied:
Thanks

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -684,11 +684,11 @@ def get_available_firmware_version(self, image_path):
 
     def update_firmware(self, image_path):
-        self.__install_firmware(image_path)
+        self.__install_firmware(image_path, reboot=True)
```
@alexrallen why need to explicitly pass this flag?

alexrallen replied:
We don't

nazariig commented:
@alexrallen a generic comment: what would happen when user has scheduled a BIOS/ONiE/SSD (Innodisk) FW update simultaneously?

alexrallen commented:
All three will schedule for during reboot. When the reboot plugin runs, it will run the BIOS / ONIE installer which will queue the two images and set the reboot flag, then it will run the SSD update and immediately do a register power cycle after that is done.

alexrallen commented:
@nazariig please take a look

nazariig commented:
> All three will schedule for during reboot. When the reboot plugin runs, it will run the BIOS / ONIE installer which will queue the two images and set the reboot flag, then it will run the SSD update and immediately do a register power cycle after that is done.

@alexrallen two things here:

1. Do you have a logic for letting SSD update to happen at the end of the entire flow (e.g., SSD power cycle will trigger BIOS/ONiE update)?
2. AFAIK, the ONiE FW Update Framework can process only one update at time, so basically either BIOS or ONiE will happen after switch reboot

nazariig commented on platform/mellanox/install-pending-fw.py:
```diff
@@ -0,0 +1,78 @@
+        key = comp["comp"]
+
+        if csp.is_modular_chassis(): 
+            _, parent_name, comp_name = key.split('/')
```
@alexrallen you can't simply split using `is_modular_chassis()` flag since modular chassis platforms can have chassis (mandatory) + modules (optional). Also, from auto update status file, you don't have a context (whether this component chassis or module related?), so looks like the only way is to parse the component path + use different maps for chassis/module components

alexrallen replied:
Then perhaps I misunderstood your initial suggestion to use this function in place of parsing the key?

alexrallen replied:
I believe I have fixed this in the latest revision although it is basically identical to the initial implementation....

nazariig replied:
> Then perhaps I misunderstood your initial suggestion to use this function in place of parsing the key?

@alexrallen your understanding was correct: the review comment was confusing - it's my bad

alexrallen commented:
> > All three will schedule for during reboot. When the reboot plugin runs, it will run the BIOS / ONIE installer which will queue the two images and set the reboot flag, then it will run the SSD update and immediately do a register power cycle after that is done.
> 
> @alexrallen two things here:
> 
>     1. Do you have a logic for letting SSD update to happen at the end of the entire flow (e.g., SSD power cycle will trigger BIOS/ONiE update)?

No. I don't see why I would need to. The SSD can be updated with the default register reboot as long as we suppress the reboot from the update script. It can also be updated from the CPLD reboot which is also a register reboot. 

> 
>     2. AFAIK, the ONiE FW Update Framework can process only one update at time, so basically either BIOS or ONiE will happen after switch reboot

According to the documentation this should be possible. I do need to test.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -321,7 +326,7 @@ def update_firmware(self, image_path):
         try:
             self.__stage_update(image_path)
-            self.__trigger_update()
+            self.__trigger_update(allow_reboot)
```
@alexrallen in case of pending updates (attempt to update ONiE/BIOS simultaneously) it will fail - you need to relax the original validation

alexrallen replied:
Will do.

alexrallen commented:
Some details from BSP call. 

>Thanks for the call
Motivation: The requirement from Microsoft is that multiple firmware be able to be installed during a single reboot.
>
>Summary:
AI2:  Oren Mano please advise with Innodisk whether we can run suppress this automatic reboot after FW upgrade without impacting existing functionality (by simply adding a –no-power-cycle option to the mlnx-ssd-fw-upgrade.sh script which just exits the script with a stern warning to reboot immediately rather than executing that reboot) and for us to continue and call some one more app (refresh) which does the power cycle ? if answer is that its dangerous what if we lock the file system right after FW upgrade will it suffice to protect ? 
>
>The conclusion regarding the modified flow would be :
>
>Option 1: CPLD update is required (SSD update might be needed too)
>Updating system components.
>apply CPLD update (without refresh).
>If relevant: apply SSD FW upgrade (without full power cycle) and right after Block SSD writes: 
>AI2:  @Alexander Allen to add this and test SSD status (blockage should be release after power cycle)
>apply CPLD Refresh.
>
>Option 2:_SSD update is required (but NOT CPLD update).
>Updating system components.
>apply SSD FW upgrade with full power cycle

nazariig commented on platform/mellanox/install-pending-fw.py:
```diff
@@ -0,0 +1,96 @@
+    if type(c) == ComponentCPLD:
+        c.install_firmware(files[i])
+    else:
+        c.install_firmware(files[i], allow_reboot=False)
```
@alexrallen what's the idea behind of two lists here? Why not using a list of tuples?

alexrallen replied:
Personal preference, I like to avoid tuples unless they are named tuples to make code more readable. I was on the fence here but I chose the two lists over a single list of named tuples because I thought it was a bit simpler.

nazariig replied:
@alexrallen understood

alexrallen commented on platform/mellanox/mlnx-ssd-fw-update.sh:
```diff
@@ -280,12 +287,12 @@ function get_ssd_info() {
+        if [ ! -x "$(command -v ${DEPENDECIES[$i]})" ]; then
+            LOG_MSG_AND_EXIT "Error: This tool require the following utils to be installed ${DEPENDECIES[$i]}"
+        fi
+    done
```
The file was incorrectly using tabs here, I switched to space (along with corrects where my editor had put in tabs).

nazariig replied:
@alexrallen understood

alexrallen commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -514,32 +504,28 @@ def auto_update_firmware(self, image_path, boot_action):
 
         # boot_type did not match (skip)
-        return FW_AUTO_ERR_BOOT_TYPE
+        if boot_action != "cold":
```
Not sure I understand your question. The branch before checks if if a reboot is required to upgrade the SSD, if a reboot is not required then it installs and returns.

If a reboot is required we need to check that a reboot has actually been specified to the automatic firmware update command, else we need to fail out because we require a reboot and an insufficient reboot was provided.

nazariig replied:
@alexrallen understood

alexrallen commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -756,6 +742,29 @@ def __install_firmware(self, image_path):
+            return FW_AUTO_ERR_BOOT_TYPE
+
+        # Install burn. Error if fail.
+        if not self.install_firmware(image_path):
```
In this case you return True or False depending on success or failure. The only other error that would be handled is a runtime error thrown by the MPFAManager class which seems to be elegant enough to let through.

nazariig replied:
@alexrallen ok. Let's have it as is.

alexrallen commented on platform/mellanox/install-pending-fw.py:
```diff
@@ -0,0 +1,96 @@
+            if not mpfa.get_metadata().has_option('firmware', 'refresh'):
+                raise RuntimeError("Failed to get {} refresh firmware".format(self.name))
+            refresh_firmware = mpfa.get_metadata().get('firmware', 'refresh')
+            comp_install = comp_install + [component]
```
Given I need to append and prepend to the list I figured this was the most stylistically correct because I could use the same pattern to do both.

nazariig replied:
@alexrallen understood

alexrallen commented on platform/mellanox/install-pending-fw.py:
```diff
@@ -0,0 +1,96 @@
+            files = [fw_file] + files
+
+# Do install
+for i, c in enumerate(comp_install):
```
This will not be possible given how the appending / pre-pending works. CPLD is always last.

nazariig replied:
@alexrallen understood

alexrallen commented on platform/mellanox/install-pending-fw.py:
```diff
@@ -0,0 +1,96 @@
+            comp_install = comp_install + [component]
+            files = files + [os.path.join(mpfa.get_path(), refresh_firmware)]
+        else:
+            comp_install = [component] + comp_install
```
This logic creates a correctly ordered list of firmware to install, I have to do this and unpack the CPLD in advance because if the SSD firmware locks the filesystem prior to install the CPLD will not be able to unpack. So I use this to "prepare" all the firmware to be able to install without needed to write to the filesystem and to also install in the correct order (SSD then CPLD).

nazariig replied:
@alexrallen understood

alexrallen commented on platform/mellanox/install-pending-fw.py:
```diff
@@ -0,0 +1,96 @@
+        if type(component) == ComponentCPLD:
+            if CPLD_FLAG:
+                print("WARNING: Multiple CPLD firmwares defined. Some CPLD updates may not fully complete.")
+            CPLD_FLAG = True
```
Will fix. Thanks.

nazariig commented on platform/mellanox/install-pending-fw.py:
```diff
@@ -0,0 +1,97 @@
+            mpfa = MPFAManager(fw_file)
+            mpfa.extract()
+            if not mpfa.get_metadata().has_option('firmware', 'refresh'):
+                raise RuntimeError("Failed to get {} refresh firmware".format(self.name))
```
@alexrallen IMHO, it's a bad idea to raise exceptions here. Who should handle it? Suggest using try-catch block for MPFA operations

alexrallen replied:
Hrm. I think I agree, I had just copied this logic. We should probably just handle the error and continue.

alexrallen replied:
Resolved.

nazariig commented on platform/mellanox/install-pending-fw.py:
```diff
@@ -0,0 +1,97 @@
+# Do install
+for i, c in enumerate(comp_install):
+    if type(c) == ComponentCPLD:
+        c.install_firmware(files[i])
```
@alexrallen how about exceptions handling for FW installation/upgrade?

alexrallen replied:
Hrm, python exceptions? This is directly before the reboot so I'm not sure how much any type of exception handling is going to help us here. 

I believe the exceptions should be handled at the implementation of `install_firmware` as necessary (which I believe is the case now). This install call will return True / False depending on the success but there isn't anything we can really act on at this point in the process. The user is going to have to let the switch reboot and check the status of the firmware.

nazariig replied:
@alexrallen there could be a backtrace so the output will be messed.

Please have a look on how it is done in `fwutil`:
```python
https://github.com/Azure/sonic-utilities/blob/master/fwutil/main.py#L209

    try:
        click.echo("Installing firmware:")
        click.echo(TAB + fw_path)
        log_helper.log_fw_install_start(component_path, fw_path)
        status = component.install_firmware(fw_path)
        log_helper.log_fw_install_end(component_path, fw_path, status)
    except KeyboardInterrupt:
        log_helper.log_fw_install_end(component_path, fw_path, False, "Keyboard interrupt")
        raise
    except Exception as e:
        log_helper.log_fw_install_end(component_path, fw_path, False, e)
        cli_abort(ctx, str(e))
```

alexrallen replied:
If the file system is locked then at this stage we will not have write access to a log. 

I am fine try-catching this and dumping an error to the terminal though if that is okay with you.

alexrallen replied:
@nazariig I have added a commit, please check.

alexrallen commented:
@nazariig Do you mind taking a look at these latest changes? Thank you.

nazariig commented on slave.mk:
```diff
@@ -939,6 +939,7 @@ $(addprefix $(TARGET_PATH)/, $(SONIC_INSTALLERS)) : $(TARGET_PATH)/% : \
                 $(LIBNSS_TACPLUS) \
                 $(MONIT) \
                 $(OPENSSH_SERVER) \
+		$(SMARTMONTOOLS) \
```
@alexrallen please remove this change and raise a new PR apart from this one. Still have a doubts that Community will accept it...

alexrallen replied:
I talked with the MSFT maintainer and they were okay with this. Was concerned as well. 

I can raise it separately.

nazariig commented on platform/mellanox/mlnx-platform-api/sonic_platform/component.py:
```diff
@@ -133,6 +135,17 @@ class ONIEUpdater(object):
 
+    BIOS_UPDATE_FILE_EXT = '.rom'
+
+    def __add_prefix(self, image_path):
```
@alexrallen what is the motivation for this change? What is the use case? IMHO, looks like a hack.

alexrallen replied:
I can't say I love it either I am more than happy to get your thoughts.

ONIE has to be installed before BIOs during boot and the installer processes the files in alphabetical order.

So I append 000 to the ONIE name so no matter what it is named it will always install first.

alexrallen commented:
@nazariig I implemented your suggestions. Also split the smartmontools work into a new PR https://github.com/alexrallen/sonic-buildimage/pull/23

Please take a look. I am checking on CI now.
